File: agent/RedByte.Agent/Network/Crypto.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using NSec.Cryptography;

namespace RedByte.Agent.Network;

public static class Crypto
{
    public const int PublicKeySize = 32;
    private const int NonceSize = 12;
    private const int TagSize = 16;

    private static byte[]? _sharedKey;

    public static byte[] CreatePublicKey(out Key privateKey)
    {
        privateKey = Key.Create(KeyAgreementAlgorithm.X25519, new KeyCreationParameters
        {
            ExportPolicy = KeyExportPolicies.AllowPlaintextExport
        });

        return privateKey.PublicKey.Export(KeyBlobFormat.RawPublicKey);
    }

    public static void SetSharedKey(Key privateKey, byte[] serverPublicKey)
    {
        if (serverPublicKey.Length != PublicKeySize)
            throw new Exception("Invalid server public key.");

        PublicKey serverKey = PublicKey.Import(KeyAgreementAlgorithm.X25519, serverPublicKey, KeyBlobFormat.RawPublicKey);
        SharedSecret? sharedSecret = KeyAgreementAlgorithm.X25519.Agree(privateKey, serverKey);
        if (sharedSecret == null)
            throw new Exception("Could not create shared key.");

        HkdfSha256 hkdf = new HkdfSha256();
        _sharedKey = hkdf.DeriveBytes(sharedSecret, null, null, 32);
    }

    public static byte[] Encrypt(object data)
    {
        byte[] key = GetSharedKey();
        byte[] nonce = RandomNumberGenerator.GetBytes(NonceSize);
        byte[] plainBytes = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(data));
        byte[] cipherBytes = new byte[plainBytes.Length];
        byte[] tag = new byte[TagSize];

        using var aes = new AesGcm(key, TagSize);
        aes.Encrypt(nonce, plainBytes, cipherBytes, tag);

        // Frame body: nonce(12) | ciphertext | tag(16)
        byte[] result = new byte[nonce.Length + cipherBytes.Length + tag.Length];
        nonce.CopyTo(result, 0);
        cipherBytes.CopyTo(result, nonce.Length);
        tag.CopyTo(result, nonce.Length + cipherBytes.Length);

        return result;
    }

    public static T? Decrypt<T>(byte[] frame)
    {
        if (frame.Length < NonceSize + TagSize)
            throw new Exception("Invalid encrypted frame.");

        byte[] key = GetSharedKey();
        byte[] nonce = frame[..NonceSize];
        byte[] cipherBytes = frame[NonceSize..^TagSize];
        byte[] tag = frame[^TagSize..];
        byte[] plainBytes = new byte[cipherBytes.Length];

        using var aes = new AesGcm(key, TagSize);
        aes.Decrypt(nonce, cipherBytes, tag, plainBytes);

        string json = Encoding.UTF8.GetString(plainBytes);
        return JsonSerializer.Deserialize<T>(json);
    }

    public static void Clear()
    {
        _sharedKey = null;
    }

    private static byte[] GetSharedKey()
    {
        if (_sharedKey == null)
            throw new Exception("Secure session was not started.");

        return _sharedKey;
    }
}

File: server/network/crypto.py
import json
import os
import socket
from typing import Any
from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import x25519
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from network.protocol import MAX_FRAME_LEN, recv_exact, recv_frame, send_frame

class Crypto:
    NONCE_SIZE = 12
    TAG_SIZE = 16
    PUBLIC_KEY_SIZE = 32
    KEY_SIZE = 32
    MIN_ENCRYPTED_BODY_SIZE = NONCE_SIZE + TAG_SIZE

    _keys: dict[socket.socket, bytes] = {}

    @staticmethod
    def start_secure_session(sock: socket.socket) -> None:
        client_public_key = Crypto._recv_public_key(sock)

        server_private_key = x25519.X25519PrivateKey.generate()
        server_public_key = server_private_key.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

        shared_secret = server_private_key.exchange(client_public_key)
        Crypto._keys[sock] = Crypto._derive_aes_key(shared_secret)
        sock.sendall(server_public_key)

    @staticmethod
    def close_secure_session(sock: socket.socket) -> None:
        Crypto._keys.pop(sock, None)

    @staticmethod
    def recv_secure(sock: socket.socket) -> dict[str, Any]:
        key = Crypto._get_key(sock)
        body = recv_frame(sock)

        if len(body) < Crypto.MIN_ENCRYPTED_BODY_SIZE or len(body) > MAX_FRAME_LEN:
            raise ValueError(f"Invalid encrypted frame length: {len(body)}")

        nonce = body[:Crypto.NONCE_SIZE]
        ciphertext_and_tag = body[Crypto.NONCE_SIZE:]

        try:
            plaintext = AESGCM(key).decrypt(nonce, ciphertext_and_tag, None)
            payload = json.loads(plaintext.decode("utf-8"))
        except InvalidTag as exc:
            raise ValueError("Invalid encrypted frame tag") from exc
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise ValueError("Invalid encrypted JSON payload") from exc

        if not isinstance(payload, dict):
            raise ValueError("Encrypted payload must be a JSON object")

        return payload

    @staticmethod
    def send_secure(sock: socket.socket, payload: dict[str, Any]) -> None:
        key = Crypto._get_key(sock)
        plaintext = json.dumps(payload, separators=(",", ":")).encode("utf-8")
        nonce = os.urandom(Crypto.NONCE_SIZE)

        # pyca/cryptography AESGCM returns ciphertext || tag.
        ciphertext_and_tag = AESGCM(key).encrypt(nonce, plaintext, None)
        body = nonce + ciphertext_and_tag

        if len(body) > MAX_FRAME_LEN:
            raise ValueError("Encrypted frame too large")

        send_frame(sock, body)

    @staticmethod
    def _recv_public_key(sock: socket.socket) -> x25519.X25519PublicKey:
        public_key_bytes = recv_exact(sock, Crypto.PUBLIC_KEY_SIZE)
        try:
            return x25519.X25519PublicKey.from_public_bytes(public_key_bytes)
        except ValueError as exc:
            raise ValueError("Invalid client public key") from exc

    @staticmethod
    def _derive_aes_key(shared_secret: bytes) -> bytes:
        return HKDF(
            algorithm=hashes.SHA256(),
            length=Crypto.KEY_SIZE,
            salt=None,
            info=None,
        ).derive(shared_secret)

    @staticmethod
    def _get_key(sock: socket.socket) -> bytes:
        key = Crypto._keys.get(sock)
        if key is None:
            raise RuntimeError("Secure session was not started")
        return key
